src/jit/optimizer.cpp

```diff
@@ -2834,365 +2829,373 @@ void Compiler::optUnrollLoops()
+            20, // FAST_CODE
+            0   // COUNT_OPT_CODE
+        };
+
+        noway_assert(ITER_LIMIT[SMALL_CODE] == 0);
+        noway_assert(ITER_LIMIT[COUNT_OPT_CODE] == 0);
+
+        unsigned iterLimit = (unsigned)ITER_LIMIT[compCodeOpt()];
 
-        for (unsigned lnum = 0; lnum < optLoopCount; lnum++)
+#ifdef DEBUG
+        if (compStressCompile(STRESS_UNROLL_LOOPS, 50))
         {
-            BasicBlock* block;
-            BasicBlock* head;
-            BasicBlock* bottom;
-
-            GenTree* loop;
-            GenTree* test;
-            GenTree* incr;
-            GenTree* phdr;
-            GenTree* init;
-
-            bool       dupCond;
-            int        lval;
-            int        lbeg;         // initial value for iterator
-            int        llim;         // limit value for iterator
-            unsigned   lvar;         // iterator lclVar #
-            int        iterInc;      // value to increment the iterator
-            genTreeOps iterOper;     // type of iterator increment (i.e. ASG_ADD, ASG_SUB, etc.)
-            var_types  iterOperType; // type result of the oper (for overflow instrs)
-            genTreeOps testOper;     // type of loop test (i.e. GT_LE, GT_GE, etc.)
-            bool       unsTest;      // Is the comparison u/int
-
-            unsigned totalIter;     // total number of iterations in the constant loop
-            unsigned loopCostSz;    // Cost is size of one iteration
-            unsigned loopFlags;     // actual lpFlags
-            unsigned requiredFlags; // required lpFlags
+            iterLimit *= 10;
+        }
+#endif
 
-            GenTree* loopList; // new stmt list of the unrolled loop
-            GenTree* loopLast;
+        static const int UNROLL_LIMIT_SZ[COUNT_OPT_CODE + 1] = {
+            300, // BLENDED_CODE
+            0,   // SMALL_CODE
+            600, // FAST_CODE
+            0    // COUNT_OPT_CODE
+        };
 
-            static const int ITER_LIMIT[COUNT_OPT_CODE + 1] = {
-                10, // BLENDED_CODE
-                0,  // SMALL_CODE
-                20, // FAST_CODE
-                0   // COUNT_OPT_CODE
-            };
+        noway_assert(UNROLL_LIMIT_SZ[SMALL_CODE] == 0);
+        noway_assert(UNROLL_LIMIT_SZ[COUNT_OPT_CODE] == 0);
 
-            noway_assert(ITER_LIMIT[SMALL_CODE] == 0);
-            noway_assert(ITER_LIMIT[COUNT_OPT_CODE] == 0);
+        int unrollLimitSz = (unsigned)UNROLL_LIMIT_SZ[compCodeOpt()];
 
-            unsigned iterLimit = (unsigned)ITER_LIMIT[compCodeOpt()];
+        loopFlags = optLoopTable[lnum].lpFlags;
+        // Check for required flags:
+        // LPFLG_DO_WHILE - required because this transform only handles loops of this form
+        // LPFLG_CONST - required because this transform only handles full unrolls
+        // LPFLG_SIMD_LIMIT - included here as a heuristic, not for correctness/structural reasons
+        requiredFlags = LPFLG_DO_WHILE | LPFLG_CONST | LPFLG_SIMD_LIMIT;
```

Comment:
This one makes more sense looking at the individual commits -- one change enables unrolling loops with branches and multiple exits, and hence removes the ONE_EXIT restriction here; the last change enables unrolling for SIMD loops and so adds that flag.